packages/ui5-test-writer/src/index.ts

```diff
@@ -0,0 +1,209 @@
+ * @param fs - an optional reference to a mem-fs editor
+ * @returns Reference to a mem-fs-editor
+ */
+export function generateOPAFiles(basePath: string, opaConfig: { scriptName?: string }, fs?: Editor): Editor {
+    if (!fs) {
+        fs = create(createStorage());
+    }
+
+    const manifest = fs.readJSON(join(basePath, 'webapp/manifest.json')) as any as Manifest;
+    const config = createConfig(manifest, opaConfig);
+
+    const rootCommonTemplateDirPath = join(__dirname, '../templates/common');
+    const rootV4TemplateDirPath = join(__dirname, '../templates/v4'); // Only v4 is supported for the time being
+    const testOutDirPath = join(basePath, 'webapp/test');
+
+    // Test files
+    fs.copy(join(rootCommonTemplateDirPath, 'testsuite.qunit.html'), join(testOutDirPath, 'testsuite.qunit.html'));
+    fs.copy(join(rootCommonTemplateDirPath, 'testsuite.qunit.js'), join(testOutDirPath, 'testsuite.qunit.js'));
+
+    // Integration (OPA) test files
+    fs.copyTpl(
+        join(rootV4TemplateDirPath, 'integration', 'opaTests.*.*'),
+        join(testOutDirPath, 'integration'),
+        config,
+        undefined,
+        {
+            globOptions: { dot: true }
+        }
+    );
+
+    // Pages files (one for each page in the app)
+    config.pages.forEach((page) => {
+        writePageObject(page, rootV4TemplateDirPath, testOutDirPath, fs!);
+    });
+
+    // OPA Journey file
+    const startPages = config.pages.filter((page) => page.isStartup).map((page) => page.targetKey);
+    const startListReportPage = config.pages.find((page) => {
+        return page.isStartup && page.template === 'ListReport';
+    });
+    const journeyParams = {
+        startPages,
+        startLR: startListReportPage?.targetKey
+    };
+    fs.copyTpl(
+        join(rootV4TemplateDirPath, 'integration/FirstJourney.js'),
+        join(testOutDirPath, `integration/${config.opaJourneyFileName}.js`),
+        journeyParams,
+        undefined,
+        {
+            globOptions: { dot: true }
+        }
+    );
+
+    return fs;
+}
+
+/**
+ * Generate a page object file for a Fiori elements for OData V4 application.
+ * Note: this doesn't modify other existing files in the webapp/test folder.
+ *
+ * @param basePath - the absolute target path where the application will be generated
+ * @param pageObjectParameters - parameters for the page
+ * @param pageObjectParameters.targetKey - the key of the target in the manifest file corresponding to the page
+ * @param fs - an optional reference to a mem-fs editor
+ * @returns Reference to a mem-fs-editor
+ */
+export function generatePageObjectFile(
+    basePath: string,
+    pageObjectParameters: { targetKey: string },
+    fs?: Editor
+): Editor {
+    if (!fs) {
+        fs = create(createStorage());
+    }
+
+    const manifest = fs.readJSON(join(basePath, 'webapp/manifest.json')) as any as Manifest;
```

Comment:
Input validation? To handle cases when the manifest can't be read/accessed or doesn't exist at all

Reply:
now throwing an exception if manifest can't be read